src/engine/CountAvailablePredicates.cpp

```diff
@@ -0,0 +1,144 @@
+                                                   size_t subjectColumnIndex) :
+  Operation(qec),
+  _subtree(subtree),
+  _subjectColumnIndex(subjectColumnIndex),
+  _predicateVarName("predicate"),
+  _countVarName("cont") { }
+
+
+// _____________________________________________________________________________
+string CountAvailablePredicates::asString(size_t indent) const {
+  std::ostringstream os;
+  for (size_t i = 0; i < indent; ++i) { os << " "; }
+  os << "COUNT_AVAILABLE_PREDICATES (col " << _subjectColumnIndex << ")\n"
+     << _subtree->asString(indent);
+  return os.str();
+}
+
+// _____________________________________________________________________________
+size_t CountAvailablePredicates::getResultWidth() const {
+  return 2;
+}
+
+// _____________________________________________________________________________
+size_t CountAvailablePredicates::resultSortedOn() const {
+  return -1;
+}
+
+
+// _____________________________________________________________________________
+void CountAvailablePredicates::setVarNames(const std::string& predicateVarName,
+                                           const std::string& countVarName) {
+  _predicateVarName = predicateVarName;
+  _countVarName = countVarName;
+}
+
+// _____________________________________________________________________________
+std::unordered_map<string, size_t>
+CountAvailablePredicates::getVariableColumns() const {
+  std::unordered_map<string, size_t> varCols;
+  varCols[_predicateVarName] = 0;
+  varCols[_countVarName] = 1;
+  return varCols;
+}
+
+// _____________________________________________________________________________
+float CountAvailablePredicates::getMultiplicity(size_t col) {
+  if (col == 0) {
+    return 1;
+  } else {
+    // TODO(florian): track the multiplicity of the counts column, if there is a benefit to that
+    return 1;
+  }
+}
+
+// _____________________________________________________________________________
+size_t CountAvailablePredicates::getSizeEstimate() {
+  // There is no easy way of computing the size estimate, but it should also
+  // not be used, as this operation should not be used withing the optimizer.
+  return _subtree->getSizeEstimate();
+}
+
+// _____________________________________________________________________________
+size_t CountAvailablePredicates::getCostEstimate() {
+  // This operation should not be used within the optimizer, and the cost
+  // should never be queried.
+  AD_THROW(ad_semsearch::Exception::NOT_YET_IMPLEMENTED,
+           "CountAvailablePredicates has no implementation for the cost "
+           "estimate determination.");
+  return 1;
+}
+
+// _____________________________________________________________________________
+void CountAvailablePredicates::computeResult(ResultTable* result) const {
+  // TODO(florian): make the loading of the has-pattern and has-relation relations cachable
+
+  result->_nofColumns = 2;
+  result->_sortedBy = 0;
```

Comment:
Returning -1 from resultSortedOn() indicates, that the result is not sorted on any column, while the 1 for the cost of the operation is not in any way connected to the operations actual cost. The exception is thrown to avoid anybody using the costEstimate of the function without realizing that the return value is a dummy value.